content/blog/2019-11-13-gep.md/index.md

```diff
@@ -0,0 +1,458 @@
+PointerSource = %tmp;
+CurrentType = [10 x i32]*;
+CurrentOp = i64 0;
+
+//Get the max offset for accessing %tmp. Since %tmp was generated with
+// %tmp = alloca [10 x i32]
+//We know that %tmp points to only 1 integer array
+NumElements = 1;
+InsertCheck(CurrentOp >= 0 && CurrentOp < NumElements); // (0 >= 0 && 0 < 1)
+//Our implementation will actually automatically omit this
+//check since it can be easily statically determined to be `true`
+
+CurrentType = NextElementType(CurrentType); //[10 x i32]
+CurrentOp = i32 %x;
+
+NumElements = 10 // retrieved from the type [10 x i32]
+InsertCheck(CurrentOp >= 0 && CurrentOp < NumElements); // (x >= 0 && x < 10)
+
+CurrentOp = <none>;
+//Done!
+```
+
+## Pointer Sizes and Tracking Allocations
+
+In the above examples, we could always tell how big our memory allocations
+were since they were allocated with static sizes. `int tmp[10]` comprises of
+two static allocations: 1) a single pointer-sized memory cell (to contain the local variable `tmp`); 2) a memory cell
+containing 10 integers (the memory pointed to by `tmp`).
+
+In many cases, the sizes of arrays may be difficult or impossible to determine at compile
+time. Consider the following snippet:
+
+```C
+int foo(int x, y) {
+    int tmp[x];
+    ... //init values in tmp
+    return tmp[y];
+}
+```
+In the corresponding LLVM code, the type of `tmp` is no longer a sized type;
+it is just `i32*`. We can no longer use types to help us determine what
+are and are not legal offsets. In this case, however, there is something
+that we can do. LLVM uses [`alloca`](https://llvm.org/docs/LangRef.html#alloca-instruction)
+instructions to allocate local variables. `alloca` takes an argument to
+determine how many elements must be allocated. If we keep track of the
+sizes of local allocations we can infer that the above code is safe if and only if:
+
+```C
+0 <= y < x //we'll assume x > 0 here
+```
+
+In our implementation, we simply keep around a map from allocaitons to their sizes.
+Additionally, we track heap allocations by scanning for function calls to `malloc`.
+This allows us to calculate maximum pointer index values for GEP instructions
+where the types are unsized. Unfortunately, this is rather imprecise since
+it tracks *exact* value dependencies and doesn't keep track of other ways
+a pointer may be passed to a GEP. For instance, spilling a value to memory
+and then re-loading it will cause our analysis to lose track of the original
+allocation. 
+
+Additionally, we run our transformation as a function pass so
+it doesn't track interprocedural allocations. The main
+reason for this limitation is that, even if we knew
+the original allocation size for all callers, we would
+have to modify the function signature to communicate
+legal index values from the caller.
+This seemed both out of scope for our current project
+and a potentially questionable design decision. Should
+a compiler pass be modifying the signatures of
+potentially every function?
```

Comment:
Indeed. For what it's worth, I think the sound (e.g., Valgrind or ASan) approach to this wouldn't require modifying function signatures—it would instead keep a gigantic hash table somewhere that maps addresses to size information. So when you try to index a pointer, the instrumentation would invoke the runtime to look up the actual address in this table to find its dynamic size. These hash table loads could be selectively eliminated when you have sufficient static information, but that would be an optimization rather than the primary mechanism as it is here.

Reply:
Good point, I'll mention that and point to the other group that implemented that style of check.